airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py

```diff
@@ -67,11 +68,30 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
     _stream_to_instance_map: Dict[str, Stream] = {}
     _slice_logger: SliceLogger = DebugSliceLogger()
 
+    def discover_error_handler(self) -> AbstractDiscoverErrorHandler:
+        return DefaultDiscoverErrorHandler()
+
     def discover(self, logger: logging.Logger, config: Mapping[str, Any]) -> AirbyteCatalog:
         """Implements the Discover operation from the Airbyte Specification.
         See https://docs.airbyte.com/understanding-airbyte/airbyte-protocol/#discover.
         """
-        streams = [stream.as_airbyte_stream() for stream in self.streams(config=config)]
+        streams = []
+        errors = []
+
+        for stream in self.streams(config=config):
+            try:
+                streams.append(stream.as_airbyte_stream())
+            except Exception as exception:
+                errors.append(exception)
+                self.discover_error_handler().handle_discover_error(logger=logger, exception=exception)
+
+        if not streams:
+            raise AirbyteTracedException(
+                internal_message=f"No streams were discovered in the source. Please check the logged errors for more information: {errors}",
+                message="No streams were discovered in the source. Please check the source configuration.",
+                failure_type=FailureType.config_error,
```

Comment:
open question: should we always raise a config error or should it be 
- file-based: config_error
- else: system_error

the reasoning being that I wouldn't expect other connectors to not have any streams discovered

Implementation wise, this might just mean adding a method that returns a boolean on the discover error handler